lib/pbench/server/api/resources/__init__.py

```diff
@@ -1290,12 +1282,12 @@ def _check_authorization(self, mode: ApiAuthorization):
                 # An unauthenticated user is never allowed to access private
                 # data nor to perform an potential mutation of data: REJECT
                 current_app.logger.warning(
-                    "Attempt to {} user {} data without login", role, username
+                    "Attempt to {} user {} data without login", role, user_id
                 )
                 raise UnauthorizedAccess(
                     authorized_user,
                     role,
-                    username,
+                    user_id,
```

Comment:
Now this is where I am really confused. In order to log a username we would have to make a privileged request to the OIDC server from the Pbench server and that means 2 things:
1. Either have a private pbench-server client configured with a service account enabled and have a `realm-admin` role assigned to it.
2. Have a user with `realm-admin` role assigned to it.

Either way, because we don't maintain the user database we will have to make a Rest API call to the OIDC server (from pbench server) to get the "admin" token and use that token on the userinfo endpoint to get the user information. (Remember this is not for the authorized user, this is for another user that the "authorized user" is trying to get access to like dataset owner id).

Is that something we want to do, its certainly possible but not sure if that's very efficient?

Reply:
Either that or we could record the username in our tables like the dataset table along with a `user_id` field when someone uploads the data so we have the information available to use to log in without accessing the OIDC server. I think this would be an ideal approach.

Reply:
@dbutenhof, why do you think it's a problem to expose user IDs?  They aren't actually secrets or even part of credentials, are they?

As for translating user IDs to usernames, this is a logistical challenge.

One way is for the Pbench Server to ask the Keycloak Broker for the translation, but, as Nikhil points out, the requires the Pbench Server to connect directly to the Broker and it requires the Pbench Server to be configured with the required credentials/authorization to access that information (and we don't know what the performance of the translation will be).

The other approach is for the Pbench Server to cache ID-username translations (e.g., in a database table) and update them periodically (e.g., each time a user presents a token).  This frees us from having to contact Keycloak for the translation, but we instead have to make a database query (which may or may not be cheaper).  And, the translation will only be as good as the last time the target user made an authenticated request of the Pbench Server (which, IMNSHO is perfectly adequate, but Dave wanted the translation to be "fresher" than that).

In any case, for the immediate future, I'm not concerned about exposing user IDs (I hope that that is not naive...).  And, I think we should follow the caching approach, because I think it is simpler, especially from a deployment perspective, and I think it will be adequate to our needs.

Reply:
> I don't want to expose them because they're not human-friendly; that's all.

Fair enough.

I also wouldn't want to report them if they aren't useful (or if reporting them just makes finding the useful information harder).  So, the question is, would including the user ID here be useful?  I'm thinking that it is not, since we're already reporting it in the immediately preceding log message.

However, this raises a stickier question:  we're instantiating an `UnauthorizedAccess` object here, and we don't have values for half of the parameters!  😆  (We've got `role` and `access` (and, of course, the HTTP status), but `authorized_user` is `None` and we would prefer to omit `user_id`.)

Perhaps we need to reexamine what an `UnauthorizedAccess` object comprises....

Reply:
> However, this raises a stickier question: we're instantiating an `UnauthorizedAccess` object here, and we don't have values for half of the parameters!

For both `UnauthorizedAccess` and `UnauthorizedAdminAccess` it would probably be good enough to say "The authenticated user isn't authorized to {munge} the specified resource". We're chattier than we need to be, and in addition the objects hold references to DB proxy objects, and we've historically had problems with `__str__` being executed when the objects are out of scope.

Reply:
These changes I'll defer in PR [Rework the User table](https://github.com/distributed-system-analysis/pbench/pull/3251), but the circular situation is that without the changes from this PR I really can not break the Users table as it will break the functional test and also unit test for which I have opened another PR to refactor the unit tests #3249. And to address this change here I have to have  [Rework the User table](https://github.com/distributed-system-analysis/pbench/pull/3251) in. 

Therefore to enable the user table changes we discussed last week I have to merge this PR and PR #3249 first to completely remove the dependency on the User table from functional and unit tests.